tests/foreman/api/test_architecture.py

```diff
@@ -14,119 +14,90 @@
 
 :Upstream: No
 """
+import pytest
+from fauxfactory import gen_choice
 from nailgun import entities
 from requests.exceptions import HTTPError
 
 from robottelo.datafactory import invalid_names_list
+from robottelo.datafactory import parametrized
 from robottelo.datafactory import valid_data_list
 from robottelo.decorators import tier1
-from robottelo.decorators import tier2
-from robottelo.decorators import upgrade
-from robottelo.test import APITestCase
 
 
-class ArchitectureTestCase(APITestCase):
-    """Tests for architectures."""
+@pytest.fixture(scope="module")
+def module_os():
+    module_os = entities.OperatingSystem().create()
+    yield module_os
+    module_os.delete()
 
-    @tier2
-    def test_positive_add_os(self):
-        """Create an architecture and associate it with an OS.
 
-        :id: 9943063d-34f3-4dbc-a341-474ec781e4d9
+@pytest.fixture(scope="module")
+def module_arch():
+    module_arch = entities.Architecture().create()
+    yield module_arch
+    module_arch.delete()
 
-        :expectedresults: The architecture can be created, and the association
-            can be read back from the server.
 
-        :CaseLevel: Integration
-        """
-        operating_sys = entities.OperatingSystem().create()
-        arch = entities.Architecture(operatingsystem=[operating_sys]).create()
-        self.assertEqual({operating_sys.id}, {os.id for os in arch.operatingsystem})
+@tier1
+def test_positive_CRUD(module_os):
+    """Create a new Architecture with several attributes, update the name
+    and delete the Architecture itself.
 
-    @tier1
-    def test_positive_create_with_name(self):
-        """Create an architecture providing the initial name.
+    :id: 80bca2c0-a6a1-4676-a036-bd918812d600
 
-        :id: acbadcda-3410-45cb-a3aa-932a0facadc1
+    :expectedresults: Architecture should be created, modified and deleted successfully
+        with given attributes.
 
-        :expectedresults: Architecture is created and contains provided name.
+    :CaseImportance: Critical
+    """
 
-        :CaseImportance: Critical
-        """
-        for name in valid_data_list().values():
-            with self.subTest(name):
-                arch = entities.Architecture(name=name).create()
-                self.assertEqual(name, arch.name)
+    # Create
+    name = gen_choice(list(valid_data_list().values()))
```

Comment:
I guess the name should be parametrized the same way as you did in `test_negative_create_with_invalid_name`

Reply:
IIRC I had a discussion with @rplevka way back, that it's probably not worth to test all the names from the valid_data_list() on every test run. 

I don't know it's possible to use the pytest.parametrize here, becuase that way the whole crud test would be executed for each item in the valid_data_list().

 A lot of these crud end to end tests in robotello do it [this way](https://github.com/SatelliteQE/robottelo/blob/602103047e677d7aeb41c777afdb94a5aa4403df/tests/foreman/api/test_computeresource_azurerm.py#L56). Some [of them](https://github.com/SatelliteQE/robottelo/blob/602103047e677d7aeb41c777afdb94a5aa4403df/tests/foreman/api/test_reporttemplates.py#L104)  loops over the list in the test. Another option would be to keep separate parametrized test just for testing entity creation with names from list.

I would suggest the option with looping over them in the test, but again I don't know if it's something that we want to do every time.

Reply:
Good point :+1:  Thanks for explanation and examples :)